### PR TITLE
flush_timeout, client perf, and downstairs DTrace

### PR DESCRIPTION
**Pull request by leftwo**
### **New CrucibleOpts field: flush_timeout**
This adds the ability to select how long the Upstairs waits before issuing
the internal flush. The ability to chose this at runtime will allow tests to
try various values and report performance or other info without having
to recompile.

### **Client perf test updates:**
Added read and write loop options to client perf test. This controls
how many of each type of IO workload we do.
Broke the read and write phases of the perf test into their own loops.

### **DTrace probes in the downstairs.**
We now have probes for when work is taken off the wire, when work
is submitted to the OS, when work returns from the OS, and when work
is ACK'd back to the upstairs (put back on the wire).

This required passing the job_id down a layer deeper than it was before,
and adding job id's where none were needed before (and are ignored).

Added some DTrace example scripts.

Here is an example from a dtrace script:
```
alan@atrium:crucible$ pfexec dtrace -s ./tools/dtrace/perf-downstairs.d         
dtrace: script './tools/dtrace/perf-downstairs.d' matched 18 probes             
^C                                                                              
                                                                                
    19316  read                                                                 
           value  ------------- Distribution ------------- count                
           16384 |                                         0                    
           32768 |@@@@@                                    1208                 
           65536 |@@@@@                                    1319                 
          131072 |@@@@                                     1071                 
          262144 |@@@@@@                                   1421                 
          524288 |@@@@@@@@                                 1880                 
         1048576 |@@@@@@                                   1458                 
         2097152 |@@@@@@@                                  1643                 
         4194304 |                                         0                    
                                                                                
    19315  read                                                                 
           value  ------------- Distribution ------------- count                
            8192 |                                         0                    
           16384 |                                         3                    
           32768 |@@@@@                                    1283                 
           65536 |@@@@@                                    1251                 
          131072 |@@@@                                     1095                 
          262144 |@@@@@@                                   1395                 
          524288 |@@@@@@@@                                 1886                 
         1048576 |@@@@@@                                   1447                 
         2097152 |@@@@@@                                   1607                 
         4194304 |                                         32                   
         8388608 |                                         1                    
        16777216 |                                         0                    
                                                                                
    19317  read                                                                 
           value  ------------- Distribution ------------- count                
           16384 |                                         0                    
           32768 |@@@                                      799                  
           65536 |@@@@@@@                                  1643                 
          131072 |@@@@@                                    1189                 
          262144 |@@@@@@                                   1405                 
          524288 |@@@@@@@                                  1874                 
         1048576 |@@@@@@                                   1394                 
         2097152 |@@@@@@                                   1593                 
         4194304 |                                         100                  
         8388608 |                                         3                    
        16777216 |                                         0                    
                                                                                
    19315  flush                                                                
           value  ------------- Distribution ------------- count                
         2097152 |                                         0                    
         4194304 |@@@                                      2                    
         8388608 |@@@@@@                                   4                    
        16777216 |@@@                                      2                    
        33554432 |                                         0                    
        67108864 |@@                                       1                    
       134217728 |                                         0                    
       268435456 |                                         0                    
       536870912 |@@@@@@                                   4                    
      1073741824 |@@@@@@@@                                 5                    
      2147483648 |@@@@@@                                   4                    
      4294967296 |@@@@@@                                   4                    
      8589934592 |                                         0                    
                                                                                
    19316  flush                                                                
           value  ------------- Distribution ------------- count                
         2097152 |                                         0                    
         4194304 |@@@                                      2                    
         8388608 |@@@@@@                                   4                    
        16777216 |@@@                                      2                    
        33554432 |@@                                       1                    
        67108864 |                                         0                    
       134217728 |                                         0                    
       268435456 |                                         0                    
       536870912 |@@@@@                                    3                    
      1073741824 |@@@@@@@@@                                6                    
      2147483648 |@@@@@@                                   4                    
      4294967296 |@@@@@@                                   4                    
      8589934592 |                                         0   

    19317  flush                                              
           value  ------------- Distribution ------------- count    
         2097152 |                                         0        
         4194304 |@@@                                      2        
         8388608 |@@@@@@                                   4        
        16777216 |@@@                                      2        
        33554432 |@@                                       1        
        67108864 |                                         0        
       134217728 |                                         0        
       268435456 |                                         0        
       536870912 |@@@@@                                    3        
      1073741824 |@@@@@@@@@                                6        
      2147483648 |@@@@@@                                   4        
      4294967296 |@@@@@@                                   4        
      8589934592 |                                         0        

    19315  write                                              
           value  ------------- Distribution ------------- count    
          262144 |                                         0        
          524288 |                                         42       
         1048576 |@@@@@@@@@@@@@@@@@@@@@@@@                 5888     
         2097152 |@@@@@@@                                  1845     
         4194304 |@@                                       559      
         8388608 |@@@@@                                    1236     
        16777216 |@                                        335      
        33554432 |                                         55       
        67108864 |                                         30       
       134217728 |                                         9        
       268435456 |                                         0        
       536870912 |                                         1        
      1073741824 |                                         0        

    19316  write                                              
           value  ------------- Distribution ------------- count    
          262144 |                                         0        
          524288 |                                         55       
         1048576 |@@@@@@@@@@@@@@@@@@@@@@                   5605     
         2097152 |@@@@@@@@                                 2069     
         4194304 |@@                                       588      
         8388608 |@@@@@                                    1238     
        16777216 |@                                        351      
        33554432 |                                         65       
        67108864 |                                         28       
       134217728 |                                         0        
       268435456 |                                         0        
       536870912 |                                         0        
      1073741824 |                                         1        
      2147483648 |                                         0        

    19317  write                                              
           value  ------------- Distribution ------------- count    
          262144 |                                         0        
          524288 |                                         24       
         1048576 |@@@@@@@@@@@@@@@@@@@@                     5074     
         2097152 |@@@@@@@@@                                2338     
         4194304 |@@@                                      739      
         8388608 |@@@@@                                    1294     
        16777216 |@                                        343      
        33554432 |                                         120      
        67108864 |                                         65       
       134217728 |                                         1        
       268435456 |                                         0        
       536870912 |                                         1        
      1073741824 |                                         0        
      2147483648 |                                         0        
      4294967296 |                                         1        
      8589934592 |                                         0       
```

I thought it would be fun to come back someday and look at those terrible times :)

### CODE REVIEW

File: client/src/main.rs [hunk 1]
```diff
@@ -1405,20 +1418,35 @@ struct Record {
...
+    write_loop: usize,
+    read_loop: usize,
 ) -> Result<()> {
+    // Before we start, make sure the work queues are empty.
+    loop {
+        let wc = guest.query_work_queue()?;
+        if wc.up_count + wc.ds_count == 0 {
+            break;
+        }
+        std::thread::sleep(std::time::Duration::from_secs(2));
```
Comment (jmpesp) [inline client/src/main.rs:1447]: I think these should be tokio::time::sleep(...).await instead? this blocks the whole thread

File: upstairs/src/lib.rs [hunk 2]
```diff
@@ -6595,8 +6597,17 @@ async fn up_listen(
     dst: Vec<Target>,
     mut ds_status_rx: mpsc::Receiver<Condition>,
     mut ds_reconcile_done_rx: mpsc::Receiver<Repair>,
+    timeout: Option<u32>,
 ) {
     println!("Wait for all three downstairs to come online");
+    let flush_timeout = {
+        if timeout.is_some() {
+            timeout.unwrap()
```
Comment (jmpesp) [inline upstairs/src/lib.rs:6605]: could use https://doc.rust-lang.org/std/option/enum.Option.html#method.unwrap_or

**Review by jmpesp — APPROVED** [review]

File: upstairs/src/lib.rs (hunk 2 above)
Comment (leftwo) [reply upstairs/src/lib.rs:6605]: nice!

File: client/src/main.rs (hunk 1 above)
Comment (leftwo) [reply client/src/main.rs]: Updated.

File: upstairs/src/lib.rs (hunk 2 above)
Comment (leftwo) [reply upstairs/src/lib.rs]: done

**Pull request merged**